python/lsst/ci/prepare.py

```diff
@@ -644,6 +655,284 @@ async def pull_worker(queue):
+                        head_ref=pr["head"]["ref"],
+                        sha=pr["head"]["sha"],
+                    )
+                    pr_items.append(pr_col)
+
+        # Model to hold all product PR items
+        pr_info = AllProductPRCollections(items=pr_items)
+
+        # Print summary
+        pr_info.summary()
+
+        manager = PRInfoManager(self.build_dir)
+        manager.save_pr_info(pr_info)
+
+        return pr_info
+
+
+class PRInfoManager:
+    """
+    Writes the Pydantic model: AllProductPRCollections to the build dir
+    and can load it back.
+
+    Parameters
+    ----------
+    build_dir : str
+        The directory in which `pr_info.json` will be written
+        and read.
+    """
+
+    def __init__(self, build_dir: str) -> None:
+        """
+        Initialize with the build directory where pr_info.json
+        will be stored.
+        """
+        self.build_dir: str = build_dir
+        self.pr_info_file: str = os.path.join(self.build_dir, "pr_info.json")
+
+    def save_pr_info(self, pr_info: AllProductPRCollections) -> None:
+        """
+        Write the AllProductPRCollections model to the build_dir
+        as pr_info.json.
+
+        Parameters
+        ----------
+        pr_info : AllProductPRCollections
+            The Pydantic model containing PR information to serialize.
+
+        Raises
+        ------
+        ValueError
+            If `self.build_dir` does not exist or is not a directory.
+        """
+        # Ensure build_dir exists
+        if not os.path.isdir(self.build_dir):
+            raise ValueError(f"build_dir {self.build_dir} does not exist or is not a directory.")
+
+        # Remove any cached pr file before start
+        if os.path.exists(self.pr_info_file):
+            os.remove(self.pr_info_file)
+
+        # Write to self.build_dir
+        with open(self.pr_info_file, "w", encoding="utf-8") as f:
+            f.write(pr_info.model_dump_json(indent=2))
+
+    def load_pr_info(self) -> AllProductPRCollections:
+        """
+        Load the PR info from pr_info.json if it exists,
+        otherwise return empty list.
+
+        Returns
+        -------
+        AllProductPRCollections
+            The Pydantic model. If `pr_info.json` does not exist or fails
+            to parse or validate, returns an empty model
+            (`AllProductPRCollections(items=[])`).
+        """
+        if not os.path.exists(self.pr_info_file):
+            print(f"pr_info.json does not exist in {self.build_dir}")
+            return AllProductPRCollections(items=[])
+        try:
+            with open(self.pr_info_file, encoding="utf-8") as f:
+                data = f.read()
+
+            pr_info = AllProductPRCollections.model_validate_json(data)
+            return pr_info
+
+        except (json.JSONDecodeError, OSError) as os_err:
+            print(f"WARNING: Could not read {self.pr_info_file}: {os_err}")
+            return AllProductPRCollections(items=[])
```

Comment:
you can move this and the later return out of the except block and only use a single return.